### DISCUSSION TITLE
How supported is the ESP32?

### DISCUSSION THREAD

**Discussion by Frogmore42** [arendst/Tasmota#10129] [Q&A] [answered]
I have a very chatty heat pump water heater that speaks 9-bit serial over RS485.  I have decoded/reverse engineered the protocol and now know what is saying and want to have Tasmota support for it.  But, the esp8266 is not really capable of 9-bit serial at a fast enough pace (115,200 baud) and keep the Wi-Fi code happy.  Will the dual cores on the ESP32 allow me to implement software serial (using interrupts) without disrupting the Wi-Fi code.

It also generates a lot of data, so it really needs more RAM than the esp8266 has.  How much free RAM/heap is there on the ESP32?
My esp8266 devices typically have 19-23KB free.  I was able to make it fit and work on the esp8266 (as a proof of concept) but the hardware watchdog would fire every 4-12 hours (so just a little bit worse than some of the not so good core versions 😉 )

**Comment by kugelkopf123** [comment, marked as answer]
ESP32 Works quite good. 
Have You tried the sml driver with script mode? There is also a raw mode in the driver. Perhaps it already has everything zu need for your heater. 
https://tasmota.github.io/docs/Smart-Meter-Interface/
And 
https://tasmota.github.io/docs/Scripting-Language/

**Reply by Frogmore42** [reply]
Thanks for the pointer.  It looks interesting.  I don't see anyone has decided the protocol from the water heater and it is not documented.  It is also 9-bit serial, with the 9-th bit set to indicate the start of a packet/datagram.  Interesting data is at particular byte positions in particular typed packets, so the general process used for other meters might work.

**Reply by kugelkopf123** [reply]
Yes. Perhaps you can give it a try with something like:
```
>D
>B
=>sensor53 r
>M 1
1,3,r,0,115200,Heater
x(@1,TEST,kwh,test,4
#
```
Then type sensor53 d1 into your webconsole to see what the driver hopefully can extract. 
1,**3**,r,0,115200,Heater
The 3 is the RX GPIO. You can choose for another HW Serial Gpio.
You have nothing to set for that within the device GPIO Settings. All this settings are done via Script.
